File: day-3/python/main.py
"""Entry point for application."""
import typing

from parser import parse
from wire import Point


def get_input() -> typing.List[str]:
    """Open input file and return lines."""
    with open("../input", "r") as f:
        return f.readlines()


if __name__ == "__main__":
    wires = [parse(inst.split(",")) for inst in get_input()]

    wire1_points = set(wires[0].points)
    wire2_points = set(wires[1].points)

    intersections = list(wire1_points.intersection(wire2_points))

    intersections.remove(Point(0, 0))

    orig = Point(0, 0)
    intersections.sort(key=lambda point: orig.manhattan_distance(point))

    part1 = intersections[0].manhattan_distance(orig)

    print(f"Part 1: {part1}")

    trip_distances = map(
        lambda p: (p, wires[0].dist_to(p) + wires[1].dist_to(p)),
        intersections
    )

    trip_distances = list(trip_distances)

    trip_distances.sort(key=lambda y: y[1])

    print(f"Part 2: {trip_distances[0][1]}")

File: day-3/rust/src/main.rs
use std::fs::File;
use std::io::Read;

mod wire;
mod parser;

use std::collections::HashSet;
use wire::{Wire, Point};

fn main() {
    let wires = get_input().expect("Could not parse wire data.");

    let sets = wires.iter().map(|wire| wire.clone().to_set()).collect::<Vec<HashSet<Point>>>();

    let a = &sets[0];
    let b = &sets[1];
    let origin = Point::new(0, 0);

    let manhattan_intersection = a.intersection(b)
        .into_iter()
        .map(|p| (p, p.manhattan_distance(&origin)))
        .collect::<Vec<(&Point, i64)>>();

    let mut smallest: (&Point, i64) = manhattan_intersection[1];

    for (p, dist) in manhattan_intersection {
        if p.x == 0 && p.y == 0 {
            continue;
        }

        if dist < smallest.1 {
            smallest = (p, dist);
        }
    }

    println!("Part 1: {}", smallest.1);

    let smallest_steps_intersection = a.intersection(b)
        .into_iter()
        .collect::<Vec<&Point>>();

    let mut first = true;
    let mut smallest: (&Point, i64) = (&origin, 0);

    for point in smallest_steps_intersection {
        if point.x == 0 && point.y == 0 {
            continue;
        }

        let mut combined = 0;

        for wire in &wires {
            combined += wire.clone().find_steps_to(point) + 1;
        }

        if first {
            smallest = (point, combined);
            first = false;
        } else {
            if combined < smallest.1 {
                smallest = (point, combined);
            }
        }
    }

    println!("Part 2: {}", smallest.1);
}

fn get_input() -> Result<Vec<Wire>, std::io::Error> {
    let mut f = File::open("../input")?;

    let mut string = String::new();

    f.read_to_string(&mut string)?;

    let wires = string.split_whitespace().map(|wire| parser::parse(wire)).collect::<Vec<Wire>>();

    Ok(wires)
}
